### PR TITLE
BUG: fix np.ma.median with only one non-masked value and an axis argument

### LINKED ISSUE #8029: Regression in numpy master: np.ma.median with axis returns wrong result when there is only one non-masked value
### Snippet:
```py
import numpy as np

masked_arr = np.ma.masked_array([[5.], [0.], [0.]],
                                [[False], [True], [True]])
print('masked_arr\n', masked_arr)
print('median\n', np.ma.median(masked_arr, axis=0))
```

#### numpy 1.11.1
```masked_arr
 [[5.0]
 [--]
 [--]]
median
 [ 5.]
```
The median is 5. as expected since this is the only non-masked value.

#### numpy master
```
masked_arr
 [[5.0]
 [--]
 [--]]
median
 [2.5]
```
The median is 2.5, unexpectedly.

FWIW, the result is fine when not using an `axis` argument.

### PR DESCRIPTION
**Pull request by lesteve**
Fix #8029.

### CODE REVIEW

File: numpy/ma/extras.py [hunk 1]
```diff
@@ -736,7 +736,7 @@ def _median(a, axis=None, out=None, overwrite_input=False):
     # duplicate high if odd number of elements so mean does nothing
     odd = counts % 2 == 1
     if asorted.ndim > 1:
-        np.copyto(low, high, where=odd)
+        low[odd] = high[odd]
```
Comment (lesteve) [inline numpy/ma/extras.py]: `np.copyto(low, high, where=odd)` doesn't unmask the values in contrary to `low[odd] = high[odd]`.

Comment (charris) [inline numpy/ma/extras.py]: Now there are indexing errors because odd is boolean...

Comment (lesteve) [inline numpy/ma/extras.py]: Not sure I am following, do you have an example where things go wrong?

Comment (charris) [inline numpy/ma/extras.py]: There were problems when odd was a scalar or scalar array. Perhaps that no longer happens with the other changes...

Comment (charris) [inline numpy/ma/extras.py]: So now that we are using the `count` function, that may be fixed. In fact, by the time we get there 0-d inputs should have been raveled to 1-d and 1-d arrays have been handled, so the `asorted.ndim > 1` check should not be needed either.

I don't think any of the values should be masked except for zero values of `count`, which brings up the question of negative indices in creating low.

Comment (charris) [inline numpy/ma/extras.py]: OK, it is the `h == 0` case that causes problems because of the wrap to masked values.

Comment (charris) [inline numpy/ma/extras.py]: The negative index also causes problems for empty arrays when `arr.shape[axis] == 0`.

Comment (charris) [inline numpy/ma/extras.py]: Could you add a note about the wrapping to masked values and the importance of unmasking those values when creating `low`?

Comment (lesteve) [inline numpy/ma/extras.py]: I am afraid I haven't got a grasp of the issue as good as yours, but what about replacing `h - 1` by `np.maximum(h - 1, 0)` ?

Comment (lesteve) [inline numpy/ma/extras.py]: This way you can keep using `np.copyto` as previously.

Comment (charris) [inline numpy/ma/extras.py]: I think that should work also, I was just a bit concerned about speed. OTOH, it is a bit more obvious. Does is work?

Comment (charris) [inline numpy/ma/extras.py]: If we do that, can probably remove `low._sharedmask = False` also.

Comment (lesteve) [inline numpy/ma/extras.py]: > I think that should work also, I was just a bit concerned about speed. OTOH, it is a bit more obvious. Does is work?

The tests pass on my machine, I pushed the change. Let's see what Travis and AppVeyor have to say!

> If we do that, can probably remove low._sharedmask = False also.

OK I can try that too afterwards.

**Comment by charris** [comment]
When (if) this passes the tests, would you squash the commits?

**Comment by lesteve** [comment]
> When (if) this passes the tests, would you squash the commits?

The tests were passing so I squashed the commits. BTW just in case you didn't know about it you can do squash + merge via the github UI these days.

**Comment by charris** [comment]
I don't really trust git to do it right after trying their revert button ;) Haven't tried squashing though.

**Comment by charris** [comment]
Thanks @lesteve. I still want to revisit `ma.median` at some point, but I think we now have something that works. Of course, I've thought that before. If nothing else, the tests need to have better coverage.

**Comment by lesteve** [comment]
Great!

> If nothing else, the tests need to have better coverage.

If you have suggestions of weird edge cases to add I'd be happy to give it a go.

**Comment by charris** [comment]
If nothing else, there should be tests for different values of `ndim`, say up to at least three, with all different valid axis values tested with different amounts of masking from all to none, and maybe empty arrays also (which will currently fail for some cases). There is also the `out` and `overwrite_input` keywords. It is a large combinational mess, but if we had had such tests we wouldn't have run into so many problems.

**Comment by lesteve** [comment]
Thanks for the suggestions, I'll try to find some time to add more tests.

**Pull request merged**